procedures/unit-testing-utils.ipf

```diff
@@ -893,68 +908,167 @@ static Function NumBytesInUTF8Character(str, byteOffset)
-							char2num("\u200E"), char2num("\u200F"), char2num("\u061C"), char2num("\u202A"), \
-							char2num("\u202D"), char2num("\u202B"), char2num("\u202E"), char2num("\u202C"), \
-							char2num("\u2066"), char2num("\u2067"), char2num("\u2068"), char2num("\u2069")}
-
-	len = strlen(str)
-	if(len)
-		do
-			numBytes = NumBytesInUTF8Character(str, bytePos)
-			code = char2num(str[bytePos, bytePos + numBytes - 1])
-			if(code >= char2num("\u0000") && code <= char2num("\u001F"))
-				// C0 controls
-				extPrint = 1
-			elseif(code >= char2num("\u007F") && code <= char2num("\u009F"))
-				// C1 controls
-				extPrint = 1
+	string strOut
+
+	strOut = EscapeString(str[max(start, diffpos - MAX_STRING_DIFF_CONTEXT), diffpos - 1])
+	strOut += EscapeString(str[diffpos, min(endpos, diffpos + MAX_STRING_DIFF_CONTEXT)])
+	return strOut
+End
+
+static Function GetTextDiffPos(str1, str2)
+	string str1, str2
+
+	variable i
+
+	for(i=0; i < strlen(str1); i += 1)
+		if (!!CmpStr(str1[i], str2[i]))
+			return i
+		endif
+	endfor
+
+	return -1
+End
+
+static Function DetectEndOfLine(str, start, marker)
+	variable start
+	string str, &marker
+
+	variable i
+
+	for(i = start; i < strlen(str); i += 1)
+		if (!CmpStr(str[i], "\r"))
+			if (!CmpStr(str[i + 1], "\n"))
+				// windows line ending
+				marker = "\r\n"
 			else
-				// Bi Directional characters
-				FindValue/V=(code) biDiChars
-				if(V_Value >= 0)
-					extPrint = 1
-				endif
+				// osx line ending
+				marker = "\r"
 			endif
-			format = "%0" + num2str(2 * numBytes) + "x"
-			sprintf hexOut, format, code
-			extOut += hexOut + " "
-			bytePos += numBytes
-		while(bytePos < len)
-	endif
+			return i
+		endif
+		if (!CmpStr(str[i], "\n"))
+			// unix line ending
+			marker = "\n"
+			return i
+		endif
+	endfor
 
-	return SelectString(extPrint, str, ": " + extOut)
+	// no line ending
+	marker = ""
+	return strlen(str)
 End
-#else
-static Function/S AutoExtendedStringOutput(str)
+
+static Function/S EscapeString(str)
 	string str
 
-	variable bytePos, code
-	variable extPrint, len
-	string format, hexOut
-	string extOut = ""
-
-	len = strlen(str)
-	for(bytePos = 0; bytePos < len; bytePos += 1)
-		code = char2num(str[bytePos]) & 0x7F
-		if(code >= 0 && code <= 31)
-			extPrint = 1
-		elseif(code == 127)
-			extPrint = 1
-		endif
-		sprintf hexOut, "%02x", code
-		extOut += hexOut + " "
+	variable i, charnum
+	string result = "", hex
+
+	for(i = 0; i < strlen(str); i += 1)
+		result += " "
+		charnum = char2num(str[i])
+		if (charnum < 0)
+			charnum += 256
+		endif
+		// normal printable characters
+		if(charnum >= 32 && charnum < 127)
+			result += str[i]
+			continue
+		endif
+		// control characters or multi-byte characters in UTF-8
+		strswitch(str[i])
+			case "\000":
+				result += "<NUL>"
+				break
+			case "\n":
+				result += "<\\n>"
+				break
+			case "\r":
+				result += "<\\r>"
+				break
+			case "\t":
+				result += "<\\t>"
+				break
+			default:
+				sprintf hex, "<%02x>", charnum
```

Comment:
Would prefixing hex numbers with 0x would have an advantage?

Would using `X` instead for the format be better readable?